tests/failpoints/cases/test_conf_change.rs

```diff
@@ -316,3 +316,52 @@ fn test_handle_conf_change_when_apply_fsm_resume_pending_state() {
         cluster.must_put(format!("kk{}", i).as_bytes(), b"v1");
     }
 }
+
+// This test simulates a scenario where a configuration change has been applied
+// on the target peer, allowing a leader transfer to that target peer even if
+// the change hasn't been applied on the current leader.
+//
+// The setup involves a 4-node cluster where peer-1 starts as the leader. A
+// configuration change is initiated to remove peer-2. This configuration
+// change commits successfully but only applies on peer-4. Meanwhile, it
+// remains pending on the leader (peer-1) and on peer-3.
+//
+// The expected result is that under these circumstances, a leader transfer
+// cannot occur to peer-3 due to its unapplied configuration change. However,
+// a transfer to peer-4 should succeed since it has already applied the
+// change.
+#[test]
+fn test_applied_conf_change_on_target_peer_allows_transfer_leader() {
+    let mut cluster = new_server_cluster(0, 4);
+    let pd_client = cluster.pd_client.clone();
+    pd_client.disable_default_operator();
+    let region_id = cluster.run_conf_change();
+    pd_client.must_add_peer(region_id, new_peer(2, 2));
+    pd_client.must_add_peer(region_id, new_peer(3, 3));
+    pd_client.must_add_peer(region_id, new_peer(4, 4));
+
+    cluster.must_put(b"k1", b"v1");
+
+    fail::cfg("apply_on_conf_change_1_1", "pause").unwrap();
+    fail::cfg("apply_on_conf_change_3_1", "pause").unwrap();
+
+    pd_client.remove_peer(region_id, new_peer(2, 2));
+    sleep_ms(300);
+    // Peer 2 is still exists since the leader hasn't applied the ConfChange
+    // yet.
+    pd_client.must_have_peer(region_id, new_peer(2, 2));
+
+    // Use async_put for insertion here to avoid timeout errors, as synchronize put
+    // would fail due to the leader's apply process being paused.
+    cluster.async_put(b"k2", b"v2");
+
+    pd_client.transfer_leader(region_id, new_peer(3, 3), vec![]);
+    pd_client.region_leader_must_not_be(region_id, new_peer(3, 3));
+
+    pd_client.transfer_leader(region_id, new_peer(4, 4), vec![]);
+    pd_client.region_leader_must_be(region_id, new_peer(4, 4));
```

Comment:
Please also test the scenario that a peer is demoting from voter to learner, and the leader tries to   transfer leadership to this peer because leader's apply is block.

Reply:
Very helpful comment!